lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -837,6 +837,65 @@ bool firrtl::areTypesWeaklyEquivalent(FIRRTLType destFType, FIRRTLType srcFType,
          destFlip == srcFlip;
 }
 
+/// Returns whether the srcType can be const-casted to the destType.
+bool firrtl::areTypesConstCastable(FIRRTLType destFType, FIRRTLType srcFType,
+                                   bool srcOuterTypeIsConst) {
+  // Identical types are always castable
+  if (destFType == srcFType)
+    return true;
+
+  auto destType = destFType.dyn_cast<FIRRTLBaseType>();
+  auto srcType = srcFType.dyn_cast<FIRRTLBaseType>();
+
+  // For non-base types, only castable if identical.
+  if (!destType || !srcType)
+    return false;
+
+  // Cannot cast non-'const' src to 'const' dest
+  if (destType.isConst() && !srcType.isConst() && !srcOuterTypeIsConst)
+    return false;
+
+  // Vector types can be casted if they have the same size and castable element
+  // type.
+  auto destVectorType = destType.dyn_cast<FVectorType>();
+  auto srcVectorType = srcType.dyn_cast<FVectorType>();
+  if (destVectorType && srcVectorType)
+    return destVectorType.getNumElements() == srcVectorType.getNumElements() &&
+           areTypesConstCastable(
+               destVectorType.getElementType(), srcVectorType.getElementType(),
+               srcOuterTypeIsConst || srcVectorType.isConst());
+
+  // Bundle types can be casted if they have the same size, element names,
+  // and castable element types.
+  auto destBundleType = destType.dyn_cast<BundleType>();
+  auto srcBundleType = srcType.dyn_cast<BundleType>();
+  if (destBundleType && srcBundleType) {
+    auto destElements = destBundleType.getElements();
+    auto srcElements = srcBundleType.getElements();
+    size_t numDestElements = destElements.size();
+    if (numDestElements != srcElements.size())
+      return false;
+
+    for (size_t i = 0; i < numDestElements; ++i) {
+      auto destElement = destElements[i];
+      auto srcElement = srcElements[i];
+      if (destElement.name != srcElement.name ||
+          destElement.isFlip != srcElement.isFlip ||
```

Comment:
Not a big deal but was curious: apparently can cast `const.bundle<a: T, const b: U>` to `const.bundle<const a: T, b: U>` which seems harmless since AFAIK the two are basically equivalent anyway (?).

Reply:
It's weird but I consider it to be ok.